### ISSUE TITLE
Bzip2 fails to build (dlls have no exports)

### ISSUE THREAD

**Issue by pravic** [Microsoft/vcpkg#682]
> warning: Import libs were not present in V:/vcpkg/packages/bzip2_x86-windows/debug/lib
> Found 4 error(s). Please correct the portfile:
    V:\vcpkg\ports\bzip2\portfile.cmake
Error: Building package bzip2:x86-windows failed with: POST_BUILD_CHECKS_FAILED

Full log:
```
T:\>vcpkg install boost
-- CURRENT_INSTALLED_DIR=V:/vcpkg/installed/x86-windows
-- DOWNLOADS=V:/vcpkg/downloads
-- CURRENT_PACKAGES_DIR=V:/vcpkg/packages/bzip2_x86-windows
-- CURRENT_BUILDTREES_DIR=V:/vcpkg/buildtrees/bzip2
-- CURRENT_PORT_DIR=V:/vcpkg/ports/bzip2/.
-- Downloading http://www.bzip.org/1.0.6/bzip2-1.0.6.tar.gz...
-- Downloading http://www.bzip.org/1.0.6/bzip2-1.0.6.tar.gz... OK
-- Testing integrity of downloaded file...
-- Testing integrity of downloaded file... OK
-- Extracting source V:/vcpkg/downloads/bzip2-1.0.6.tar.gz
-- Extracting done
-- Applying patch V:/vcpkg/ports/bzip2/fix-import-export-macros.patch
-- Applying patch failed. This is expected if this patch was previously applied.
-- Applying patch V:/vcpkg/ports/bzip2/fix-import-export-macros.patch done
-- Configuring x86-windows-rel
-- Configuring x86-windows-rel done
-- Configuring x86-windows-dbg
-- Configuring x86-windows-dbg done
-- Package x86-windows-rel
-- Package x86-windows-rel done
-- Package x86-windows-dbg
-- Package x86-windows-dbg done
-- Applying patch V:/vcpkg/ports/bzip2/auto-define-import-macro.patch
-- Applying patch failed. This is expected if this patch was previously applied.
-- Applying patch V:/vcpkg/ports/bzip2/auto-define-import-macro.patch done
-- Performing post-build validation
Import libs were not present in V:/vcpkg/packages/bzip2_x86-windows/debug/lib
If this is intended, add the following line in the portfile:
    SET(VCPKG_POLICY_DLLS_WITHOUT_LIBS disabled)
Import libs were not present in V:/vcpkg/packages/bzip2_x86-windows/lib
If this is intended, add the following line in the portfile:
    SET(VCPKG_POLICY_DLLS_WITHOUT_LIBS disabled)
The following DLLs have no exports:

    V:/vcpkg/packages/bzip2_x86-windows/debug/bin/libbz2.dll
    V:/vcpkg/packages/bzip2_x86-windows/bin/libbz2.dll

DLLs without any exports are likely a bug in the build script.
There should be no empty directories in V:/vcpkg/packages/bzip2_x86-windows
The following empty directories were found:

    V:/vcpkg/packages/bzip2_x86-windows/debug/lib
    V:/vcpkg/packages/bzip2_x86-windows/lib

If a directory should be populated but is not, this might indicate an error in the portfile.
If the directories are not needed and their creation cannot be disabled, use something like this in the portfile to remove them:

    file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/a/dir ${CURRENT_PACKAGES_DIR}/some/other/dir)


-- Performing post-build validation done
Found 4 error(s). Please correct the portfile:
    V:\vcpkg\ports\bzip2\portfile.cmake
Error: Building package bzip2:x86-windows failed with: POST_BUILD_CHECKS_FAILED
Please ensure sure you're using the latest portfiles with `vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: bzip2:x86-windows
  Vcpkg version: 0.0.74-unknownhash

Additionally, attach any relevant sections from the log files above.
```

**Comment by ras0219-msft** [comment]
Hi @pravic,

It looks like bzip2 failed to build. Could you attach the log files:
```
V:/vcpkg/buildtrees/bzip2/config-x86-windows-rel-out.log
V:/vcpkg/buildtrees/bzip2/config-x86-windows-rel-err.log
V:/vcpkg/buildtrees/bzip2/package-x86-windows-rel-err.log
V:/vcpkg/buildtrees/bzip2/package-x86-windows-rel-out.log
```

**Comment by codicodi** [comment]
```
...
-- Downloading http://www.bzip.org/1.0.6/bzip2-1.0.6.tar.gz...
...
-- Applying patch V:/vcpkg/ports/bzip2/fix-import-export-macros.patch
-- Applying patch failed...
...
```
This patch should have worked on freshly downloaded sources. Maybe there are some issues with your copy of git. These logs may also help:
```
V:/vcpkg/buildtrees/bzip2/patch-x64-windows-0-err.log
V:/vcpkg/buildtrees/bzip2/patch-x64-windows-0-out.log
```

**Comment by pravic** [comment]
bzip2 compilation ended without errors.

config-x86-windows-dbg-err:
```
CMake Warning:
  Manually-specified variables were not used by the project:

    Boost_COMPILER
    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY
    CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY
```

config-x86-windows-dbg-out:
```
-- The C compiler identification is MSVC 19.0.24215.1
-- The CXX compiler identification is MSVC 19.0.24215.1
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: V:/vcpkg/buildtrees/bzip2/x86-windows-dbg 
```

patch-x86-windows-0-err is empty.

**Comment by pravic** [comment]
libbz2,dll was compiled successfully. And yes, it contains no export, as my first log said (_DLLs without any exports are likely a bug in the build script._).

**Comment by ras0219-msft** [comment]
I'd like to make sure that the patch [1] applied correctly to your sources. Could you run
```
>type buildtrees\bzip2\src\bzip2-1.0.6\bzlib.h | findstr "BZ_API"
```
from a command prompt in your Vcpkg directory?

If there is any mention of `WINAPI` in the output, the patch wasn't applied. In that case, please try deleting `buildtrees\bzip2\` and doing a fresh install of bzip2.

If this fails, please post the patch logs (both `-out` AND `-err`) again, because they should explain why the patch didn't apply.

[1] https://github.com/Microsoft/vcpkg/blob/master/ports/bzip2/fix-import-export-macros.patch

**Comment by pravic** [comment]
Well, I found what's wrong.

`vcpkg_apply_patches` unable to find `git` program, because I use `git.cmd` (no idea how `find_program` works - in ideal world it should not care about program extension, it just should try to execute any of PATHEXT). And since noone checks `find_program` result, cmake tries to operate with `GIT-NOTFOUND` program to apply patches.

**Comment by pravic** [comment]
There is a bug about that: https://cmake.org/Bug/bug_relationship_graph.php?bug_id=9879.
It seems, it is not fixed still: https://gitlab.kitware.com/cmake/cmake/issues/9879

**Comment by pravic** [comment]
Resolved by #698, works well now. Thanks for help!
